server/src/fstar_connection.ts

```diff
@@ -0,0 +1,512 @@
+			// Only keep track of responses if we expect a response
+			this.pending_responses.set(qid, {ip: internal_promise, is_stream});
+		}
+		// TODO(klinvill): Should we always return a promise even if no response
+		// is expected? In that case the promise is just ignored.
+		return internal_promise.promise;
+	}
+
+	// Send a request to F* to check the given code.
+	async fullBufferRequest(code: string, kind: 'full' | 'lax' | 'cache' | 'reload-deps', withSymbols: boolean, debug: boolean): partialResult<IdeProgress> {
+		if (!this.fstar.supportsFullBuffer) {
+			throw new Error("ERROR: F* process does not support full-buffer queries");
+		}
+		const query: FullBufferQuery = {
+			query: "full-buffer",
+			args: {
+				kind,
+				"with-symbols": withSymbols,
+				code: code,
+				line: 0,
+				column: 0
+			}
+		};
+
+		const expect_response = true;
+		const is_stream = true;
+		return this.request(query, debug, expect_response, is_stream);
+	}
+
+	// Send a request to F* to check the given code up through a position.
+	//
+	// TODO(klinvill): Since this FStarConnection object is only for one F*
+	// process (lax or not), do we need the `kind` argument here or can we infer
+	// it from the F* process?
+	async partialBufferRequest(code: string, kind: 'verify-to-position' | 'lax-to-position', position: { line: number, column: number }, debug: boolean): partialResult<IdeProgress> {
+		if (!this.fstar.supportsFullBuffer) {
+			throw new Error("ERROR: F* process does not support full-buffer queries");
+		}
+		const query: FullBufferQuery = {
+			query: "full-buffer",
+			args: {
+				kind,
+				"with-symbols": false,
+				code: code,
+				line: 0,
+				column: 0,
+				"to-position": position
+			}
+		};
+
+		const expect_response = true;
+		const is_stream = true;
+		return this.request(query, debug, expect_response, is_stream);
+	}
+
+	// Look up information about an identifier in a given file.
+	//
+	// For more details, see:
+	// https://github.com/FStarLang/FStar/wiki/Editor-support-for-F*#lookup
+	async lookupQuery(filePath: string, position: Position, word: string, range: FStarRange, debug: boolean): Promise<IdeSymbol> {
+		const query: LookupQuery = {
+			query: "lookup",
+			args: {
+				context: "code",
+				symbol: word,
+				"requested-info": ["type", "documentation", "defined-at"],
+				location: {
+					filename: filePath,
+					line: position.line + 1,
+					column: position.character
+				},
+				"symbol-range": range
+			}
+		};
+		return this.request(query, debug);
+	}
+
+	// Request to map a file into F*'s virtual file system.
+	//
+	// For more details, see:
+	// https://github.com/FStarLang/FStar/wiki/Editor-support-for-F*#vfs-add
+	async vfsAddRequest(filePath: string, contents: string, debug: boolean) {
+		const query: VfsAdd = {
+			query: "vfs-add",
+			args: {
+				filename: filePath,
+				contents: contents
+			}
+		};
+
+		const expect_response = false;
+		this.request(query, debug, expect_response);
+	}
```

Comment:
I would introduce a separate function for simple requests that have exactly one response, to reduce the Boolean blindness here.

```suggestion
        async simpleRequest<T>(query: T) {
                const expectResponse = true;
                return this.request(query, expectResponse);
        }

	// Request to map a file into F*'s virtual file system.
	//
	// For more details, see:
	// https://github.com/FStarLang/FStar/wiki/Editor-support-for-F*#vfs-add
	async vfsAddRequest(filePath: string, contents: string, debug: boolean) {
	        return this.simpleRequest<VfsAdd>({
			query: "vfs-add",
			args: {
				filename: filePath,
				contents: contents,
			},
		});
	}
```